File: design-patterns-ou-padroes-de-projeto/creational-patterns/factory-method/factory-method.py
class Car:
    def drive(self):
        print("Dirigindo um carro 🚗")

class Bike:
    def ride(self):
        print("Pedalando uma bicicleta 🚲")

class VehicleFactory:
    @staticmethod
    def create(type_):
        if type_ == "car":
            return Car()
        elif type_ == "bike":
            return Bike()
        else:
            raise ValueError("Tipo de veículo desconhecido")

# Uso da Factory
vehicle1 = VehicleFactory.create("car")
vehicle2 = VehicleFactory.create("bike")

vehicle1.drive()  # "Dirigindo um carro 🚗"
vehicle2.ride()   # "Pedalando uma bicicleta 🚲"

File: design-patterns-ou-padroes-de-projeto/creational-patterns/factory-method/factory-method.cs
using System;

public class Car
{
    public void Drive()
    {
        Console.WriteLine("Dirigindo um carro 🚗");
    }
}

public class Bike
{
    public void Ride()
    {
        Console.WriteLine("Pedalando uma bicicleta 🚲");
    }
}

public static class VehicleFactory
{
    public static object Create(string type)
    {
        return type switch
        {
            "car" => new Car(),
            "bike" => new Bike(),
            _ => throw new ArgumentException("Tipo de veículo desconhecido")
        };
    }
}

// Uso da Factory
class Program
{
    static void Main()
    {
        var vehicle1 = VehicleFactory.Create("car") as Car;
        var vehicle2 = VehicleFactory.Create("bike") as Bike;

        vehicle1?.Drive();  // "Dirigindo um carro 🚗"
        vehicle2?.Ride();   // "Pedalando uma bicicleta 🚲"
    }
}
